[ADR-001] bookmark manager

robsdedude opened the pull request:
To do:
 - [x] test against https://github.com/neo4j-drivers/testkit/pull/502
 - [x] test against https://github.com/neo4j-drivers/testkit/pull/506
 - [x] write unit tests

robsdedude committed:
Initial draft of bookmark manager

robsdedude committed:
Fix unit tests for config

robsdedude committed:
TestKit backend support for bookmark manager

robsdedude committed:
Fix bookmark manager

 * respect database reported in summary
 * fix lock usage
 * other small fixes

robsdedude committed:
Add support for `ignore_bookmark_manager` session config option

robsdedude committed:
Lock bookmark manager per db

robsdedude committed:
Add `forget` method to bookmark manager

robsdedude committed:
Add API docs for the bookmark manager

robsdedude committed:
Implement bookmark_consumer and remove must_included_databases

Implement ADR adjustments to bookmark manager
https://github.com/neo-technology/drivers-adr/pull/30/

robsdedude committed:
Make Bookmarks not iterable

robsdedude committed:
Merge branch '5.0' into bookmark-manager

robsdedude committed:
TestKit backend support for BMM extension functions

robsdedude committed:
Fix corner-case usages of the BMM within Session

robsdedude committed:
Unit tests for BMM

robsdedude committed:
Fix TestKit backend handling of BMM config options

robsdedude committed:
Rename bookmark manager hooks

`bookmark_consumer` -> `bookmarks_consumer`
`bookmark_supplier` -> `bookmarks_supplier`

https://github.com/neo-technology/drivers-adr/pull/31

robsdedude committed:
TestKit add Optimization:MinimalBookmarksSet flag

robsdedude committed:
Add performance warning for when enabling the BMM

robsdedude committed:
Merge branch '5.0' into bookmark-manager

bigmontz commented on bin/make-unasync:
```diff
@@ -154,12 +154,12 @@ class CustomRule(unasync.Rule):
                     end += 1
                 else:
-                    out += self._unasync_prefix(name[start:(end - 1)])
+                    out += self._unasync_name(name[start:(end - 1)])
```
Out of curiosity, what does it change?

robsdedude replied:
See https://github.com/neo4j/neo4j-python-driver/pull/771#discussion_r951319852

bigmontz commented on docs/source/api.rst:
```diff
@@ -138,7 +139,7 @@ Alternatively, one of the auth token helper functions can be used.
 Driver
 ******
 
-Every Neo4j-backed application will require a :class:`neo4j.Driver` object.
+Every Neo4j-backed application will require a driver object.
```
Why was the class reference removed?

robsdedude replied:
Because there also is ``:class:`neo4j.AysncDriver` ``.

bigmontz replied:
The API seems to be documented using `sync` driver in most of time. For example, the `neo4j.Driver` is referred at line 145.

fbiville commented on bin/make-unasync:
```diff
@@ -154,12 +154,12 @@ class CustomRule(unasync.Rule):
                     end += 1
                 else:
-                    out += self._unasync_prefix(name[start:(end - 1)])
+                    out += self._unasync_name(name[start:(end - 1)])
```
was sind diese?

robsdedude replied:
Have a look here https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/bin/make-unasync#L186

I patched the unasync lib to also unasync strings. This is necessary to be able to also transpile doc strings. But I made the mistake to call `_unasync_prefix` which turns classes like `class AsyncFoo` into `class Foo` and similar, but does not take extra configured replacements into consideration. I had to fix it in this PR to make lines this [this one](https://github.com/neo4j/neo4j-python-driver/pull/771/files#diff-8511a91786723e653bbd0c1068ba6f888dc44b7c976ff9f9abe96005984674f9R214) transpile correctly.

fbiville commented on docs/source/api.rst:
```diff
@@ -417,6 +419,29 @@ Specify the client agent name.
+``bookmark_manager``
+--------------------
+Specify a bookmark manager for the driver to use. If present, the bookmark
+manger is used to keep all work on the driver causally consistent.
```
```suggestion
manager is used to keep all work on the driver causally consistent.
```
were you hungry when you wrote this? ;D

fbiville replied:
nitpicking: it will be causally consistent iff one does not write/use bad BM implementations

robsdedude replied:
:yum: 

Do you have a suggestion for a better formulation that doesn't just sound... unnecessarily technically correct?

fbiville replied:
maybe `causally consistent at best`? :D
it's a nit really, feel free to ignore it

fbiville commented on docs/source/api.rst:
```diff
@@ -825,7 +850,7 @@ Returning a live result object would prevent the driver from correctly managing
 This function will receive a :class:`neo4j.ManagedTransaction` object as its first parameter.
 
-.. autoclass:: neo4j.ManagedTransaction
+.. autoclass:: neo4j.ManagedTransaction()
```
why the parentheses here? `GraphDatabase` does not have them

robsdedude replied:
That tells the docs tool (sphinx) to not include the signature of the class. That's because the user is not supposed to instantiate this class themselves. Therefore, I don't want to document the constructor.

fbiville commented on docs/source/async_api.rst:
```diff
@@ -505,7 +534,7 @@ Returning a live result object would prevent the driver from correctly managing
 This function will receive a :class:`neo4j.AsyncManagedTransaction` object as its first parameter.
 
-.. autoclass:: neo4j.AsyncManagedTransaction
+.. autoclass:: neo4j.AsyncManagedTransaction()
```
same parenthesis question

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+)
+
+
+T_BmSupplier = t.Callable[[t.Optional[str]],
```
oh, you used Optional to avoid have an interface with 2 methods or having 2 callables, is that right?

robsdedude replied:
Exactly. The `None` db means "give me all the bookmarks you have". This alternative implementation is also mentioned in the [ADR (internal link)](https://github.com/neo-technology/drivers-adr/blob/12c201dfef753728a74305bd46e999fcd72c355e/adrs/001-bookmark-manager.md?plain=1#L208).

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+
+
+T_BmSupplier = t.Callable[[t.Optional[str]],
+                          t.Union[Bookmarks, t.Awaitable[Bookmarks]]]
```
since you're in the async bookmark manager, shouldn't it always return `t.Awaitable[Bookmarks]`? why the union here?

robsdedude replied:
You'd think so :D I thought I'd give users the option to pass me either an async supplier or a sync one, should they not need any IO. See 

https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/docs/source/async_api.rst?plain=1#L157-L158

you can also configure the async driver with an `AsyncBMM` or a `BMM`. I also went down this route with the custom resolver function: it *can* be async (when using the async driver), but doesn't have to.

fbiville replied:
I see, you want to be friends with users, don't you? ;D

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+
+T_BmSupplier = t.Callable[[t.Optional[str]],
+                          t.Union[Bookmarks, t.Awaitable[Bookmarks]]]
+T_BmConsumer = t.Callable[[str, Bookmarks], t.Union[None, t.Awaitable[None]]]
```
same question as above

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+        new_bms = set(new_bookmarks)
+        prev_bms = set(previous_bookmarks)
+        with self._lock:
+            if not new_bms:
```
ah, I did not even check in Go that the new bookmarks could be empty :|

robsdedude replied:
You should do this. Either in the BMM or in the session.

https://github.com/neo-technology/drivers-adr/blob/fd77505fed6eba5d3f7f5123588619fdc780a101/adrs/001-bookmark-manager.md?plain=1#L294

fbiville replied:
https://github.com/neo4j/neo4j-go-driver/pull/395 🥳

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+            if not new_bms:
+                return
+            curr_bms = self._bookmarks[database]
+            curr_bms.difference_update(prev_bms)
```
could `difference_update` just be called `remove` or `remove_all`?

robsdedude replied:
Yes, that's what it does. But the set implementation in Python [1] generally uses terminology closer to mathematical sets (difference, union, ...).

1: https://docs.python.org/3/library/stdtypes.html#set

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+            extra_bms = await AsyncUtil.callback(
+                self._bookmarks_supplier, database
+            )
+            bms.update(extra_bms.raw_values)
```
does this call alter the bookmarks stored in the `self._bookmarks`? In Go, it would, so I explicitly copy the bookmarks before the supplied ones get added to them

robsdedude replied:
Yes, it would too. Python sets are mutable objects. But https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/neo4j/_async/bookmark_manager.py#L85 is exacly that: a copy constructor, as you'd call it in other languages. That's also why this it the only instruction being guarded by the lock.

fbiville replied:
ah, missed that, thanks!

fbiville commented on neo4j/_async/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+    ) -> None:
+        new_bms = set(new_bookmarks)
+        prev_bms = set(previous_bookmarks)
+        with self._lock:
```
so you went with a simple lock, right? not a rwlock, it seems, or?

robsdedude replied:
Yes, a simple plain lock<sup>*)<sup>.

<sup>*)</sup> The lock is actually a noop lock in the async implementation, because there is not `await` while the lock is acquired: this means there is no possibility of concurrency given that Python's async implements cooperative concurrency. I.e., you never get interrupted, unless you yield to the event loop (some `await` calls).

Furthermore, there is no builtin rwlock in Python. So I would've had to roll my own solution. That definitely seems overkill for the BMM.

fbiville commented on neo4j/_async/driver.py:
```diff
@@ -62,11 +64,16 @@
+    """Accessor for :class:`neo4j.AsyncDriver` construction.
     """
 
     if t.TYPE_CHECKING:
```
oh, you have to explicitly check for this?

robsdedude replied:
You *can* make the distinction, you don't have to. But since the driver already had these config classes about that parse kwarg dicts, I figured the least intrusive thing to do would be to only explicitly state the available config options to the type checker (most IDEs should also be able to pick this up for auto completion), but continue using `**kwargs` and the config classes at runtime.

fbiville commented on neo4j/_async/driver.py:
```diff
@@ -206,6 +215,61 @@ def driver(cls, uri, *, auth=None, **config) -> AsyncDriver:
+            )
+
+        :param initial_bookmarks:
+            The initial set of bookmarks. The default bookmark manager will
```
```suggestion
            The initial set of bookmarks. The built-in implementation of the bookmark manager will
```
just saying "The default bookmark manager" may leave people to think it's enabled by default.
The default bookmark manager is currently a no-op.

fbiville commented on neo4j/_async/driver.py:
```diff
@@ -343,13 +407,16 @@ def session(
             initial_retry_delay: float = ...,
             retry_delay_multiplier: float = ...,
             retry_delay_jitter_factor: float = ...,
+            bookmark_manager: t.Union[AsyncBookmarkManager,
```
why is the bookmark manager exposed here and not at the driver level?

fbiville replied:
ah, maybe it's not the public API to create a session from the driver, or is it?

robsdedude replied:
This is the config option (i.e., `neo4j.AsyncGraphDatabase.driver(..., bookmark_manager=...)`). This can only go here or at the session level and the ADR specs it to go here.

But you are still right, that it's `neo4j.AsyncGraphDatabase.bookmark_manager(**some_config)` and `neo4j.GraphDatabase.bookmark_manager(...)` whereas it's `neo4j.bookmark_manager` in JS and Go, and porbably others. But since this methods needs to be able to return an async or sync BMM, depending on what driver you want to configure, I can't put it at the top level, unless I want to go `neo4j.bookmark_manager(...)` and `neo4j.async_bookmark_manager(...)`. Which would've been another valid option, I suppose. I'm indecisive as of which option would be better. My initial thoughts were to put it next to the driver factory method as it's functionality is pretty much coupled to the driver.

fbiville commented on neo4j/_async_compat/concurrency.py:
```diff
@@ -423,6 +436,8 @@ def notify_all(self):
+Condition: te.TypeAlias = threading.Condition
+CooperativeLock: te.TypeAlias = threading.Lock
+Lock: te.TypeAlias = threading.Lock
+CooperativeRLock: te.TypeAlias = threading.RLock
```
you're not using this or the one below, or did I miss something?

robsdedude replied:
`AsyncCooperativeRLock` (and with it `CooperativeRLock`) are actually being used in the Pool implementation
https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/neo4j/_async/io/_pool.py#L81
https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/neo4j/_sync/io/_pool.py#L81

As is the `(Async)RLock`
https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/neo4j/_async/io/_pool.py#L453
https://github.com/neo4j/neo4j-python-driver/blob/bd5e6a97cd3737f6c4cbe1503d3cc5ccc50fa0ab/neo4j/_sync/io/_pool.py#L453

fbiville commented on tests/unit/async_/io/test_class_bolt4x3.py:
```diff
@@ -230,7 +230,7 @@ async def test_hint_recv_timeout_seconds(
                                packer_cls=AsyncBolt4x3.PACKER_CLS,
                                unpacker_cls=AsyncBolt4x3.UNPACKER_CLS)
-    sockets.client.settimeout = mocker.AsyncMock()
+    sockets.client.settimeout = mocker.Mock()
```
shouldn't it remain an `AsyncMock` since we're in an async unit test?

robsdedude replied:
Nope. That's because the method is a sync one (non-blocking) even for our async socket wrapper. Here's the definition: https://github.com/neo4j/neo4j-python-driver/blob/cc76ac5a2bb6c5933e757adee5f2415a831c78ff/neo4j/_async_compat/network/_bolt_socket.py#L143

fbiville commented on tests/unit/async_/test_bookmark_manager.py:
```diff
@@ -0,0 +1,295 @@
+    assert set(await bmm.get_bookmarks(db)) == set()
+
+
+@pytest.mark.parametrize("db", ("db1", "db2", "db3"))
```
magic

robsdedude replied:
pytest's parametrization feature is incredible mighty :clap:

fbiville commented on tests/unit/async_/test_driver.py:
```diff
@@ -136,7 +140,7 @@ async def test_routing_driver_constructor(protocol, host, port, params, auth_tok
         ),
         (
-            {"ssl_context": ssl.SSLContext(ssl.PROTOCOL_TLSv1)},
+            {"ssl_context": ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)},
```
this seems unrelated?

robsdedude replied:
Caught :see_no_evil: It gets rid of a deprecation warning https://docs.python.org/3/library/ssl.html#ssl.PROTOCOL_TLSv1
I'll put it into it's own PR.

robsdedude replied:
https://github.com/neo4j/neo4j-python-driver/pull/780

fbiville commented on tests/unit/async_/test_driver.py:
```diff
@@ -184,36 +190,38 @@ def test_driver_trust_config_error(
-                                           autospec=True)
-        tx_begin_mock = mocker.patch.object(AsyncTransaction, "_begin",
-                                            autospec=True)
+        # acquire_mock = mocker.patch.object(session._pool, "acquire",
```
was ist los hier?

fbiville commented on tests/unit/async_/test_driver.py:
```diff
@@ -281,11 +291,161 @@ async def test_verify_connectivity_parameters_are_deprecated(
+
+
+@mark_async_test
+async def test_with_default_bookmark_manager(mocker) -> None:
```
```suggestion
async def test_with_builtin_bookmark_manager(mocker) -> None:
```

bigmontz commented on docs/source/async_api.rst:
```diff
@@ -149,8 +151,13 @@ Async Driver Configuration
 :class:`neo4j.AsyncDriver` is configured exactly like :class:`neo4j.Driver`
-(see :ref:`driver-configuration-ref`). The only difference is that the async
-driver accepts an async custom resolver function:
+(see :ref:`driver-configuration-ref`). The only differences are:
+* the async driver accepts an async custom resolver function,
+see :ref:`async-resolver-ref`.
+* the async driver accepts accepts either a :class:`neo4j.api.BookmarkManager`
+object or a :class:`neo4j.api.AsyncBookmarkManager` as bookmark manager.
+see :ref:`async-bookmark-manager-ref`.
```
Nice touch

robsdedude replied:
Thanks :bow:

bigmontz commented on neo4j/_sync/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+        self, database: str, previous_bookmarks: t.Iterable[str],
+        new_bookmarks: t.Iterable[str]
+    ) -> None:
+        new_bms = set(new_bookmarks)
+        prev_bms = set(previous_bookmarks)
```
Do you need to create new sets here? We may let the methods `difference_update ` and `update` handle it.

robsdedude replied:
The `Iterable` type doesn't provide a `__bool__()` method. So I can't check if its empty or not without turning it into a `set`, `list`, `tuple`, or such first. So at least `new_bms` needs the conversion. `prev_bms` might not need it. But in Python there is something called generators. It's an iterable that, when consumed, can't be consumed again. This wouldn't break here, as we're only using `prev_bms` once in this method. But I felt like going the safe route.

Alternatively, I suppose, I could upgrade the required type to [`t.Collection`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Collection). That's allow me to check if the length is 0.

bigmontz commented on neo4j/_async/work/workspace.py:
```diff
@@ -77,6 +82,76 @@ def _set_cached_database(self, database):
+                            "iterable of raw bookmarks (deprecated).")
+        self._initial_bookmarks = self._bookmarks = prepared_bookmarks
+
+    async def _get_bookmarks(self, database):
+        if self._bookmark_manager is not None:
+            # For 4.3- support: the server will not send the resolved home
+            # database back. To avoid confusion between `None` as in "all
+            # database" and `None` as in "home database" we re-write the
+            # home database to `""`, which otherwise is an invalid database
+            # name.
+            if database is None:
+                database = ""
+            self._bookmarks = tuple({
+                *await AsyncUtil.callback(
+                    self._bookmark_manager.get_bookmarks, database
+                ),
+                *self._initial_bookmarks
+            })
+        return self._bookmarks
+
+    async def _get_all_bookmarks(self):
+        if self._bookmark_manager is not None:
+            self._bookmarks = tuple({
+                *await AsyncUtil.callback(
+                    self._bookmark_manager.get_all_bookmarks,
+                ),
+                *self._initial_bookmarks
+            })
+        return self._bookmarks
```
Changing the value of `self._bookmarks` breaks the `session.last_bookmarks` contract. For instance, if you run a read query and then try to get the `session.last_bookmarks`, you will get all the bookmarks in the `BookmarkManager` instead of the bookmarks from the session.

robsdedude replied:
:scream:

bigmontz reviewed (changes requested):
The changes on the workspace might broke `session.last_bookmarks()` current behaviour.

fbiville committed:
It's manager... not manger, duh!

Signed-off-by: Rouven Bauer <rouven.bauer@neo4j.com>

robsdedude committed:
Revert unrelated change to tests

The change was meant to get rid of a deprecation warning.
It'll be moved into it's own PR.

fbiville committed:
Clean-up

 * Avoid calling the BMM implementation that's shipped with the driver "default
   bookmark manger". This could lead users to believe it's enabled by default.
 * little oversights

Signed-off-by: Rouven Bauer <rouven.bauer@neo4j.com>

robsdedude commented on neo4j/_sync/bookmark_manager.py:
```diff
@@ -0,0 +1,108 @@
+        new_bms = set(new_bookmarks)
+        prev_bms = set(previous_bookmarks)
+        with self._lock:
+            if not new_bms:
+                return
```
No need to acquire the lock for this :sweat_smile:

bigmontz committed:
Clean-up

 * Adjust BMM interface: bookmark types are now `Collection`s instead of
   `Iterable`s. This allows for slightly more efficient BMM implementations
   without creating significant friction in the usage of BMMs.
 * Docs typo.

Signed-off-by: Rouven Bauer <rouven.bauer@neo4j.com>

robsdedude committed:
Fix session.last_bookmarks() leaking bookmarks from BMM

Discovered by Antonio

robsdedude committed:
Merge branch '5.0' into bookmark-manager

robsdedude committed:
Move BMM config to session level + mark experimental

robsdedude committed:
TestKit: support session-level BMM config

https://github.com/neo4j-drivers/testkit/pull/509

robsdedude committed:
Code-style: expand abbreviations

fbiville commented on neo4j/_async/driver.py:
```diff
@@ -225,19 +229,32 @@ def bookmark_manager(
-        bookmark manager implementation so that all work is automatically
-        causally chained (i.e., all reads can observe all previous writes
-        even in a clustered setup)::
+        Basic usage example to configure sessions with the builtin bookmark
```
```suggestion
        Basic usage example to configure sessions with the built-in bookmark
```

fbiville commented on neo4j/_async/driver.py:
```diff
@@ -225,19 +229,32 @@ def bookmark_manager(
-        even in a clustered setup)::
+        Basic usage example to configure sessions with the builtin bookmark
+        manager implementation so that all work is automatically causally
+        chained (i.e., all reads can observe all previous writes even in a
```
causal consistency makes only sense in the context of clusters anyway, or?

robsdedude replied:
Yes. That's why I say "even in" because you'll get this anyway if you're not working with a cluster. What do you suggest instead?

fbiville replied:
I'd simply remove the "even" bit. Afaik, consistency models do not apply to single instances (they're supposed to be consistent with themselves else it's just... broken)

fbiville replied:
misread the sentence, sorry

fbiville commented on neo4j/_sync/driver.py:
```diff
@@ -224,19 +228,32 @@ def bookmark_manager(
-        bookmark manager implementation so that all work is automatically
-        causally chained (i.e., all reads can observe all previous writes
-        even in a clustered setup)::
+        Basic usage example to configure sessions with the builtin bookmark
```
```suggestion
        Basic usage example to configure sessions with the built-in bookmark
```

robsdedude committed:
Fix spelling

Co-authored-by: Florent Biville <445792+fbiville@users.noreply.github.com>

robsdedude committed:
Merge branch '5.0' into bookmark-manager

robsdedude committed:
Merge branch '5.0' into bookmark-manager

bigmontz reviewed (approved):
🔖 🧑‍💼

robsdedude commented:
Waiting for https://github.com/neo4j-drivers/testkit/pull/509 to be merged. Then re-run CI.

robsdedude committed:
Merge branch '5.0' into bookmark-manager
